# **Lab 2: Indexing and Subsetting (20 points)**

#### Dylan Watson


---

**IMPORTANT**:
- Before starting the lab, save a copy to your google drive by navigating to `File > Save a copy in Drive` <br>
- Then rename the copied file `Lab#_LASTNAME_FIRSTNAME.ipynb` where # is the lab assignment number, LASTNAME is your last name, and FIRSTNAME is your first name.

<br>


**Instructions**: Complete the lab below.

- <font color=red>Tasks</font> indicated in red must be completed to receive credit
- Include all code for your work
- Include markdown cells with written answers explaining your work when prompted
- **NOTE: Your completed lab must be submitted to Canvas as a .ipynb file by the end of the lab session**
---

<br>

<br>






####<font color=red>Organized</font> (1 points)
* Answers should be organized in the following format for organization and readability

<br>

```
# 1.
This will be the code block to answer the first part of a task
```

1. This will be the text answer/explanation (if prompted) for the first task

<br>

```
# 2.
This will be the code block to answer the second part of a task in a separate code cell
```

2. This will be the text answer/explanation (if prompted) for the second task

<br>

Continue this format for the remaining subtasks


<br>

---

<br>






####<font color=red>Task 1</font>  (1 points)

The code below loads and preprocesses the US Census data described in lecture
* Run the code cell
* View the first few rows using `head()` to ensure the data has been properly reformatted for analysis


In [1]:
# create url to census data
census_url <- url("https://www2.census.gov/programs-surveys/popest/tables/2010-2011/state/totals/nst-est2011-01.csv")

# load the census data into memory
census <- read.csv(census_url)

census
# remove unnecessary columns and rows
nb_row <- nrow(census)
census <- census[-c(1:8, (nb_row - 6):nb_row), -(6:10)]

# overwrite the column names using meaningful column names
names(census) <- c("state", "april2010census", "april2010base", "july2010pop", "july2011pop")

# remove periods from state names
census$state <- gsub('\\.', '', census$state)

# remove commas from populations and convert to numeric
census$april2010census <- as.numeric(gsub(',', '', census$april2010census))
census$april2010base <- as.numeric(gsub(',', '', census$april2010base))
census$july2010pop <- as.numeric(gsub(',', '', census$july2010pop))
census$july2011pop <- as.numeric(gsub(',', '', census$july2011pop))

# remove existing row names
rownames(census) <- NULL

table.with.row.headers.in.column.A.and.column.headers.in.rows.3.through.4...leading.dots.indicate.sub.parts.                                                                                                                                                                                                                                                      
1  Table 1. Annual Estimates of the Population for the United States, Regions, States, and Puerto Rico:  April 1, 2010 to July 1, 2011                                                                                                                                                                                                                               
2  Geographic Area                                                                                                                                                                                                                                                                                                                                                   
3                                                                                                                                                                                                                                                                                                                                                                    
4  United States                                                                                                                                                                                                                                                                                                                                                     
5  Northeast                                                                                                                                                                                                                                                                                                                                                         
6  Midwest                                                                                                                                                                                                                                                                                                                                                           
7  South                                                                                                                                                                                                                                                                                                                                                             
8  West                                                                                                                                                                                                                                                                                                                                                              
9  .Alabama                                                                                                                                                                                                                                                                                                                                                          
10 .Alaska                                                                                                                                                                                                                                                                                                                                                           
11 .Arizona                                                      

In [2]:
nb_row

[1] 66

In [3]:
# View the first few rows
head(census)

state      april2010census april2010base july2010pop july2011pop
1 Alabama     4779736         4779735       4785401     4802740   
2 Alaska       710231          710231        714146      722718   
3 Arizona     6392017         6392013       6413158     6482505   
4 Arkansas    2915918         2915921       2921588     2937979   
5 California 37253956        37253956      37338198    37691912   
6 Colorado    5029196         5029196       5047692     5116796

<br>

---

<br>






####<font color=red>Task 2</font>  (4 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. What is the average population across states in July 2010?
2. What is the average population across states in July 2011?
3. Was there an increase or decrease in the average population?

Hint: Use `mean()`

In [4]:
mean(census$july2010pop)
mean(census$july2011pop)

[1] 6065298

[1] 6109645

The mean pop of all states in the July 2010 and July 2011 censuses is 6.065 million and 6.109 million respectively. This shows that there was an increase in the mean population of all US states from the years 2010 to 2011.

<br>

---

<br>





####<font color=red>Task 3</font>  (2 points)
Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.


1. What is the largest population in July 2010? [Hint: Use `max()`]

The function `which.max()` provides you with the index of the maximum. For example, the output of `which.max(c(99, 256, 48))` is 2 since the second entry of the vector, 256, is the maximum.
2. Use `which.max()` to find the state with the largest population.

In [5]:
max(census$july2010pop)
census$state[which.max(census$july2010pop)]

[1] 37338198

[1] "California"

The max population of the July 2010 census is the state of California with a pop value of 37.33 million.

<br>

---

<br>






####<font color=red>Task 4</font>  (2 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.
1. Use `which.min()` to find the state with the smallest population in July 2010.
2. What is the population in this state?

In [6]:
min(census$july2010pop)
census$state[which.min(census$july2010pop)]

[1] 564554

[1] "Wyoming"

The min population of the July 2010 census is the state of Wyoming with a pop value of 564,554.

<br>

---

<br>






####<font color=red>Task 5</font>  (4 points)

The function `order()` returns the indexes of a sorted vector. For example, the output of `order(c(99, 256, 48))` is the vector `(3, 1, 2)` since the third number is the smallest (48), the first number is next (99), and the second number is the largest (256).

The example below shows how to sort a single vector:
```
x <- c(99, 256, 48)
x[order(x)]
```
<br>



1. Sort the dataframe `census` by population in July 2010 from smallest to largest and store as another dataframe called `census_sorted` <br>
Hint: `census[order(census$july2010pop), ]`
2. Show the first few lines of `census_sorted` to ensure it is correct
3. Inferring from the sorted dataframe, which areas of the US have the smallest populations? Why?
4. Inferring from the sorted dataframe, which areas of the US have the largest populations? Why?

In [7]:
census_sorted = census[order(census$july2010pop), ]
head(census_sorted)
tail(census_sorted)

state                april2010census april2010base july2010pop july2011pop
51 Wyoming              563626          563626        564554      568158     
9  District of Columbia 601723          601723        604912      617996     
46 Vermont              625741          625741        625909      626431     
35 North Dakota         672591          672591        674629      683932     
2  Alaska               710231          710231        714146      722718     
42 South Dakota         814180          814180        816598      824082

state        april2010census april2010base july2010pop july2011pop
39 Pennsylvania 12702379        12702379      12717722    12742886   
14 Illinois     12830632        12830632      12841980    12869257   
10 Florida      18801310        18801311      18838613    19057542   
33 New York     19378102        19378104      19395206    19465197   
44 Texas        25145561        25145561      25253466    25674681   
5  California   37253956        37253956      37338198    37691912

The states with the lowest populations are Wyoming, DC, Vermont, North Dakota, Alaska, and South Dakota. The states with the largest populations are California, Texas, New York, Florida, Illinois, Pennsylvania. From this we can observe that the states with the greatest populations are usually coastal states as natural resources and trade allow for large population centers and economies. The states with the lowest populations are usually interior land locked states or smaller states with fewer population centers and less trade.

<br>

---

<br>






####<font color=red>Task 6</font>  (2 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Create a dataframe of states with populations between 1 and 2 million people in July 2010 and store as `census_1to2`.
2. Among states with populations between 1 and 2 million people, which state has the largest population? Use R code to answer this question. [Hint: `which.max()`]

Hint: `census$july2010pop > 1000000 & census$july2010pop < 2000000`

In [8]:
census_1to2 = census[census$july2010pop > 1000000 & census$july2010pop < 2000000, ]
census_1to2
census_1to2$state[which.max(census_1to2$july2010pop)]

state         april2010census april2010base july2010pop july2011pop
12 Hawaii        1360301         1360301       1363359     1374810    
13 Idaho         1567582         1567582       1571102     1584985    
20 Maine         1328361         1328361       1327379     1328188    
28 Nebraska      1826341         1826341       1830141     1842641    
30 New Hampshire 1316470         1316472       1316807     1318194    
40 Rhode Island  1052567         1052567       1052528     1051302    
49 West Virginia 1852994         1852996       1854368     1855364

[1] "West Virginia"

The states with a population between 1 and 2 million in the July 2010 census are Hawaii, Idaho, Maine, Nebraska, New Hampshire, Rhode Island, and West Virginia. Among those states, the state with the greatest population is West Virginia at 1.854 million.

<br>

---

<br>






####<font color=red>Task 7</font>  (4 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Calculate the difference in population for each state between 2010 and 2011. [Hint: population 2011 - population 2010]
2. Store as another variable called `pop_diff` in the `census` dataframe
3. How many states decreased in population between 2010 and 2011? (i.e. How many states have `pop_diff` less than 0?)
4. How many states had population increases less than 10,000 between 2010 and 2011? Hint: `pop_diff` greater than 0 and less than 10,000


In [9]:
census$pop_diff = census$july2011pop - census$july2010pop
head(census)
sum(census$pop_diff < 0)
sum(census$pop_diff > 0 & census$pop_diff < 10000)

state      april2010census april2010base july2010pop july2011pop pop_diff
1 Alabama     4779736         4779735       4785401     4802740     17339  
2 Alaska       710231          710231        714146      722718      8572  
3 Arizona     6392017         6392013       6413158     6482505     69347  
4 Arkansas    2915918         2915921       2921588     2937979     16391  
5 California 37253956        37253956      37338198    37691912    353714  
6 Colorado    5029196         5029196       5047692     5116796     69104

[1] 2

[1] 13

There are 2 states with population differences less than 0 and 13 states with population differences between 0 and 10,000.

In [10]:
# below is scratch

options(repr.matrix.max.rows=Inf, repr.matrix.max.cols=Inf)

dat <- data.frame(
  ID = 1:100,  # A column with IDs from 1 to 100
  Value1 = rnorm(100),  # A column with 100 random values from a normal distribution
  Value2 = runif(100)   # A column with 100 random values from a uniform distribution
)
dat[] <- lapply(dat, function(x) seq_along(x))

# Displaying the first few rows of the dataframe
dat

ID  Value1 Value2
1     1   1      1   
2     2   2      2   
3     3   3      3   
4     4   4      4   
5     5   5      5   
6     6   6      6   
7     7   7      7   
8     8   8      8   
9     9   9      9   
10   10  10     10   
11   11  11     11   
12   12  12     12   
13   13  13     13   
14   14  14     14   
15   15  15     15   
16   16  16     16   
17   17  17     17   
18   18  18     18   
19   19  19     19   
20   20  20     20   
21   21  21     21   
22   22  22     22   
23   23  23     23   
24   24  24     24   
25   25  25     25   
26   26  26     26   
27   27  27     27   
28   28  28     28   
29   29  29     29   
30   30  30     30   
31   31  31     31   
32   32  32     32   
33   33  33     33   
34   34  34     34   
35   35  35     35   
36   36  36     36   
37   37  37     37   
38   38  38     38   
39   39  39     39   
40   40  40     40   
41   41  41     41   
42   42  42     42   
43   43  43     43   
44   44  44     44   
45   45  45     45   
46   46  46     46   
47   47  47     47   
48   48  48     48   
49   49  49     49   
50   50  50     50   
51   51  51     51   
52   52  52     52   
53   53  53     53   
54   54  54     54   
55   55  55     55   
56   56  56     56   
57   57  57     57   
58   58  58     58   
59   59  59     59   
60   60  60     60   
61   61  61     61   
62   62  62     62   
63   63  63     63   
64   64  64     64   
65   65  65     65   
66   66  66     66   
67   67  67     67   
68   68  68     68   
69   69  69     69   
70   70  70     70   
71   71  71     71   
72   72  72     72   
73   73  73     73   
74   74  74     74   
75   75  75     75   
76   76  76     76   
77   77  77     77   
78   78  78     78   
79   79  79     79   
80   80  80     80   
81   81  81     81   
82   82  82     82   
83   83  83     83   
84   84  84     84   
85   85  85     85   
86   86  86     86   
87   87  87     87   
88   88  88     88   
89   89  89     89   
90   90  90     90   
91   91  91     91   
92   92  92     92   
93   93  93     93   
94   94  94     94   
95   95  95     95   
96   96  96     96   
97   97  97     97   
98   98  98     98   
99   99  99     99   
100 100 100    100

In [11]:
dat = dat[, -( (nrow(dat) / 2):(nrow(dat) / 2 + 1) )]
dat

ID  Value1 Value2
1     1   1      1   
2     2   2      2   
3     3   3      3   
4     4   4      4   
5     5   5      5   
6     6   6      6   
7     7   7      7   
8     8   8      8   
9     9   9      9   
10   10  10     10   
11   11  11     11   
12   12  12     12   
13   13  13     13   
14   14  14     14   
15   15  15     15   
16   16  16     16   
17   17  17     17   
18   18  18     18   
19   19  19     19   
20   20  20     20   
21   21  21     21   
22   22  22     22   
23   23  23     23   
24   24  24     24   
25   25  25     25   
26   26  26     26   
27   27  27     27   
28   28  28     28   
29   29  29     29   
30   30  30     30   
31   31  31     31   
32   32  32     32   
33   33  33     33   
34   34  34     34   
35   35  35     35   
36   36  36     36   
37   37  37     37   
38   38  38     38   
39   39  39     39   
40   40  40     40   
41   41  41     41   
42   42  42     42   
43   43  43     43   
44   44  44     44   
45   45  45     45   
46   46  46     46   
47   47  47     47   
48   48  48     48   
49   49  49     49   
50   50  50     50   
51   51  51     51   
52   52  52     52   
53   53  53     53   
54   54  54     54   
55   55  55     55   
56   56  56     56   
57   57  57     57   
58   58  58     58   
59   59  59     59   
60   60  60     60   
61   61  61     61   
62   62  62     62   
63   63  63     63   
64   64  64     64   
65   65  65     65   
66   66  66     66   
67   67  67     67   
68   68  68     68   
69   69  69     69   
70   70  70     70   
71   71  71     71   
72   72  72     72   
73   73  73     73   
74   74  74     74   
75   75  75     75   
76   76  76     76   
77   77  77     77   
78   78  78     78   
79   79  79     79   
80   80  80     80   
81   81  81     81   
82   82  82     82   
83   83  83     83   
84   84  84     84   
85   85  85     85   
86   86  86     86   
87   87  87     87   
88   88  88     88   
89   89  89     89   
90   90  90     90   
91   91  91     91   
92   92  92     92   
93   93  93     93   
94   94  94     94   
95   95  95     95   
96   96  96     96   
97   97  97     97   
98   98  98     98   
99   99  99     99   
100 100 100    100